File: packages/runar-compiler/src/passes/bindingvariant-directive.ts
/**
 * `@bindingVariant` directive parsing.
 *
 * A public method may carry a `/** @bindingVariant <VARIANT> *\/` comment
 * directive selecting which Any-S OP_PUSH_TX preimage-binding construction its
 * auto-injected covenant emits:
 *
 *   - `lowS` (default): s = lowS((z + 1) mod n) against the C=1 key. Canonical
 *     s ≤ n/2, accepted under the LOW_S rule (nVersion = 1). Safe under all tx
 *     versions. Absent directive ⇒ this construction.
 *   - `all`: s = z + 1 as-is (no mod-n, no low-S) against the same C=1 key —
 *     ~45 bytes smaller. REJECTED under LOW_S; ONLY valid for spends with
 *     nVersion != 1. The author is responsible for that guarantee.
 *
 * Reuses the exact two-surface directive shape #123 established for `@sighash`
 * (JSDoc block + leading trivia): the *detection* lives in the parser, this
 * module owns the *value grammar*. Like `@sighash`, it is a TypeScript-surface
 * directive — the eight non-TS surface parsers fail closed on it.
 */

import type { BindingVariant } from './oppushtx-codegen.js';

/** The construction a method uses when it carries no `@bindingVariant` directive. */
export const BINDING_VARIANT_DEFAULT: BindingVariant = 'lowS';

/** Recognised variant names (as written in the directive), mapped to their value. */
const VARIANT_VALUES: Record<string, BindingVariant> = {
  lowS: 'lowS',
  all: 'all',
};

export type BindingVariantParseResult =
  | { variant: BindingVariant }
  | { error: string };

/**
 * Parse the value of a `@bindingVariant` directive.
 *
 * `variantText` is the raw text following `@bindingVariant` (e.g. `"all"`), with
 * any trailing comment punctuation already stripped by the caller. The match is
 * case-sensitive: `all` and `lowS` are the only accepted spellings, so a typo
 * (`All`, `low_s`, `high`) is rejected rather than silently defaulted — a
 * mis-declared binding is an exploit class (a covenant that ships `all` while its
 * spends use nVersion = 1 is unspendable; the inverse ships needless bytes).
 */
export function parseBindingVariant(variantText: string): BindingVariantParseResult {
  const raw = variantText.trim();
  if (raw.length === 0) {
    return { error: '@bindingVariant directive requires a value (`@bindingVariant all` or `@bindingVariant lowS`)' };
  }
  const variant = VARIANT_VALUES[raw];
  if (variant === undefined) {
    return { error: `@bindingVariant: unknown variant "${raw}" (valid: lowS, all)` };
  }
  return { variant };
}

/**
 * Extract and parse a `@bindingVariant` directive from a block of comment text.
 * Returns `null` when no `@bindingVariant` token is present, otherwise the parse
 * result (variant or error). Used by the parser after it has collected a
 * method's JSDoc / leading-comment trivia.
 *
 * Only a JSDoc/line-comment TAG at the start of a comment line is a directive.
 * Mid-sentence mentions (`Do NOT use @bindingVariant all`) and trailing junk
 * (`@bindingVariant: all`, `@bindingVariant all;`) are errors, not silent
 * default-to-lowS and not silent opt-in to `all`.
 */
const BINDING_VARIANT_TOKEN_RE = /@bindingVariant\b/;

function stripCommentLinePrefix(raw: string): string {
  return raw
    .replace(/^[ \t]+/, '')
    .replace(/^(?:\/\/|\/\*\*?|\*)[ \t]*/, '')
    .replace(/[ \t]*(?:\*\/)?[ \t]*$/, '');
}

export function extractBindingVariantDirective(commentText: string): BindingVariantParseResult | null {
  if (!BINDING_VARIANT_TOKEN_RE.test(commentText)) return null;
  for (const raw of commentText.split(/\r?\n/)) {
    const line = stripCommentLinePrefix(raw);
    if (!line.startsWith('@bindingVariant')) continue;
    const rest = line.slice('@bindingVariant'.length);
    if (rest.length > 0 && /[A-Za-z0-9_]/.test(rest[0]!)) continue;
    if (rest.length > 0 && rest[0] !== ' ' && rest[0] !== '\t') {
      return { error: '@bindingVariant must be a JSDoc tag at the start of a comment line (`@bindingVariant all` or `@bindingVariant lowS`)' };
    }
    return parseBindingVariant(rest.trim());
  }
  return {
    error: '@bindingVariant must be a JSDoc tag at the start of a comment line (`@bindingVariant all` or `@bindingVariant lowS`)',
  };
}

File: compilers/java/src/main/java/runar/compiler/passes/BindingVariantDirective.java
package runar.compiler.passes;

import java.util.Set;
import java.util.regex.Pattern;

/**
 * {@code @bindingVariant} directive parsing — Java port of
 * {@code packages/runar-compiler/src/passes/bindingvariant-directive.ts}
 * (and the Go {@code frontend/bindingvariant_directive.go}).
 *
 * <p>A public method may carry a {@code /** @bindingVariant <VARIANT> *&#47;}
 * comment directive selecting which Any-S OP_PUSH_TX preimage-binding
 * construction its auto-injected covenant emits: {@code "lowS"} (default; low-S
 * fixup, safe under the LOW_S rule that applies to nVersion = 1) or {@code
 * "all"} (compact non-low-S, ~45 bytes smaller, valid only for spends with
 * nVersion != 1).
 *
 * <p>The default (no directive) is {@code "lowS"} — byte-identical to the
 * historically-pinned binding blob, so existing fixtures see ZERO change.
 * Detection lives in the parser; this class owns the variant grammar.
 */
public final class BindingVariantDirective {
    private BindingVariantDirective() {}

    /** The construction used when no directive is present. */
    public static final String DEFAULT = "lowS";

    /** The recognised variant names (as written). Case-sensitive. */
    private static final Set<String> VALUES = Set.of("lowS", "all");

    /** Parse result: exactly one of {@code value} / {@code error} is non-null. */
    public record Result(String value, String error) {
        static Result ok(String v) { return new Result(v, null); }
        static Result err(String e) { return new Result(null, e); }
        public boolean isError() { return error != null; }
    }

    /**
     * Parse the value text of a {@code @bindingVariant} directive.
     *
     * <p>Case-sensitive: only {@code "lowS"} and {@code "all"} are accepted, so a
     * typo is rejected rather than silently defaulted (a mis-declared binding is
     * an exploit class).
     */
    public static Result parseBindingVariant(String variantText) {
        String raw = variantText == null ? "" : variantText.trim();
        if (raw.isEmpty()) {
            return Result.err(
                "@bindingVariant directive requires a value "
                + "(`@bindingVariant all` or `@bindingVariant lowS`)");
        }
        if (!VALUES.contains(raw)) {
            return Result.err(
                "@bindingVariant: unknown variant \"" + raw + "\" (valid: lowS, all)");
        }
        return Result.ok(raw);
    }

    private static final Pattern BINDING_VARIANT_TOKEN =
        Pattern.compile("@bindingVariant\\b");
    private static final String LINE_START_ERR =
        "@bindingVariant must be a JSDoc tag at the start of a comment line "
        + "(`@bindingVariant all` or `@bindingVariant lowS`)";

    private static boolean isIdentChar(char c) {
        return (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9') || c == '_';
    }

    private static String stripCommentLine(String raw) {
        String s = raw.replaceFirst("^[ \\t]+", "");
        if (s.startsWith("//")) s = s.substring(2).replaceFirst("^[ \\t]+", "");
        else if (s.startsWith("/**")) s = s.substring(3).replaceFirst("^[ \\t]+", "");
        else if (s.startsWith("/*")) s = s.substring(2).replaceFirst("^[ \\t]+", "");
        else if (s.startsWith("*")) s = s.substring(1).replaceFirst("^[ \\t]+", "");
        s = s.replaceFirst("[ \\t]+$", "");
        if (s.endsWith("*/")) s = s.substring(0, s.length() - 2).replaceFirst("[ \\t]+$", "");
        return s;
    }

    /**
     * Extract and parse a {@code @bindingVariant} directive from a block of
     * comment text. Returns {@code null} when no {@code @bindingVariant} token is
     * present, otherwise the parse result (value or error). Only a JSDoc/line
     * tag at the start of a comment line is a directive.
     */
    public static Result extractBindingVariantDirective(String commentText) {
        if (commentText == null || !BINDING_VARIANT_TOKEN.matcher(commentText).find()) {
            return null;
        }
        for (String raw : commentText.split("\\R", -1)) {
            String line = stripCommentLine(raw);
            if (!line.startsWith("@bindingVariant")) continue;
            String rest = line.substring("@bindingVariant".length());
            if (!rest.isEmpty() && isIdentChar(rest.charAt(0))) continue;
            if (!rest.isEmpty() && rest.charAt(0) != ' ' && rest.charAt(0) != '\t') {
                return Result.err(LINE_START_ERR);
            }
            return parseBindingVariant(rest.trim());
        }
        return Result.err(LINE_START_ERR);
    }
}
